pkg/bloomcompactor/bloomcompactor.go

```diff
@@ -106,141 +173,263 @@ func (c *Compactor) stopping(_ error) error {
+	if err != nil {
+		level.Info(util_log.Logger).Log("putting blocks to storage", err)
+		return
+	}
+
+	// Build and upload meta.json to storage
 	meta := bloomshipper.Meta{
 		// After successful compaction there should be no tombstones
 		Tombstones: make([]bloomshipper.BlockRef, 0),
-		Blocks:     []bloomshipper.BlockRef{blockRef},
+		Blocks:     []bloomshipper.BlockRef{blocks.BlockRef},
 	}
 
-	err = c.bloomShipperClient.PutMeta(ctx, meta)
-	if err != nil {
-		return err
-	}
-	_, err = c.bloomShipperClient.PutBlocks(ctx, blocks)
+	err = bloomShipperClient.PutMeta(ctx, meta)
 	if err != nil {
-		return err
+		level.Info(util_log.Logger).Log("putting meta.json to storage", err)
+		return
 	}
-	// TODO may need to change return value of this func
+
 	return nil
 }
 
-func (c *Compactor) runCompact(ctx context.Context) error {
-	// TODO set MaxLookBackPeriod to Max ingester accepts
-	maxLookBackPeriod := c.cfg.MaxLookBackPeriod
-
-	stFp, endFp := NoopGetFingerprintRange()
-	tenantID := NoopGetUserID()
+func (c *Compactor) runCompact(ctx context.Context, bloomShipperClient bloomshipper.Client, storeClient storeClient) error {
 
-	end := time.Now().UTC().UnixMilli()
-	start := end - maxLookBackPeriod.Milliseconds()
+	series, err := listSeries(ctx, storeClient)
 
-	metaSearchParams := bloomshipper.MetaSearchParams{
-		TenantID:       tenantID,
-		MinFingerprint: stFp,
-		MaxFingerprint: endFp,
-		StartTimestamp: start,
-		EndTimestamp:   end,
-	}
-
-	metas, err := c.bloomShipperClient.GetMetas(ctx, metaSearchParams)
 	if err != nil {
 		return err
 	}
 
-	if len(metas) == 0 {
-		// run compaction from scratch
-		tempDst := os.TempDir()
-		err = c.compactNewChunks(ctx, tempDst)
-		if err != nil {
-			return err
-		}
-	} else {
-		// part 2
-		// When already compacted metas exists
-		// Deduplicate index paths
-		uniqueIndexPaths := make(map[string]struct{})
-
-		for _, meta := range metas {
-			for _, blockRef := range meta.Blocks {
-				uniqueIndexPaths[blockRef.IndexPath] = struct{}{}
+	for _, s := range series {
+		err := storeClient.indexShipper.ForEach(ctx, s.tableName, s.tenant, func(isMultiTenantIndex bool, idx shipperindex.Index) error {
+			if isMultiTenantIndex {
+				return nil
 			}
-		}
 
-		// TODO complete part 2 - discuss with Owen - add part to compare chunks and blocks.
-		// 1. for each period at hand, get TSDB table indexes for given fp range
-		// 2. Check blocks for given uniqueIndexPaths and TSDBindexes
-		//	if bloomBlock refs are a superset (covers TSDBIndexes plus more outside of range)
-		//	create a new meta.json file, tombstone unused index/block paths.
-
-		// else if: there are TSDBindexes that are not covered in bloomBlocks (a subset)
-		// then call compactNewChunks on them and create a new meta.json
-
-		// else: all good, no compaction
+			_ = idx.(*tsdb.TSDBFile).Index.(*tsdb.TSDBIndex).ForSeries(
+				ctx,
+				nil,              // TODO no shards for now, revisit with ring work
```

Comment:
I think shard arg refers to the query-level sharding. IIUC we want to process all shards here so I think this TODO can be removed.